src/locale/kbd-model-map

```diff
@@ -1,72 +1,72 @@
-la-latin1		latam	pc105		-		terminate:ctrl_alt_bksp
-us			us	pc105+inet	-		terminate:ctrl_alt_bksp
-ko			kr	pc105		-		terminate:ctrl_alt_bksp
-ro-std			ro	pc105		std		terminate:ctrl_alt_bksp
-de-latin1		de	pc105		-		terminate:ctrl_alt_bksp
-slovene			si	pc105		-		terminate:ctrl_alt_bksp
-hu101			hu	pc105		qwerty		terminate:ctrl_alt_bksp
-jp106			jp	jp106		-		terminate:ctrl_alt_bksp
-croat			hr	pc105		-		terminate:ctrl_alt_bksp
-it2			it	pc105		-		terminate:ctrl_alt_bksp
-hu			hu	pc105		-		terminate:ctrl_alt_bksp
-sr-latin		rs	pc105		latin		terminate:ctrl_alt_bksp
-fi			fi	pc105		-		terminate:ctrl_alt_bksp
-fr_CH			ch	pc105		fr		terminate:ctrl_alt_bksp
-dk-latin1		dk	pc105		-		terminate:ctrl_alt_bksp
-fr			fr	pc105		-		terminate:ctrl_alt_bksp
-it			it	pc105		-		terminate:ctrl_alt_bksp
-ua-utf			ua,us	pc105		-		terminate:ctrl_alt_bksp,grp:shifts_toggle,grp_led:scroll
-fr-latin1		fr	pc105		-		terminate:ctrl_alt_bksp
-sg-latin1		ch	pc105		de_nodeadkeys	terminate:ctrl_alt_bksp
 be-latin1		be	pc105		-		terminate:ctrl_alt_bksp
-dk			dk	pc105		-		terminate:ctrl_alt_bksp
-fr-pc			fr	pc105		-		terminate:ctrl_alt_bksp
+bg_bds-utf8		bg,us	pc105		-		terminate:ctrl_alt_bksp,grp:shifts_toggle,grp_led:scroll
 bg_pho-utf8		bg,us	pc105		,phonetic	terminate:ctrl_alt_bksp,grp:shifts_toggle,grp_led:scroll
-it-ibm			it	pc105		-		terminate:ctrl_alt_bksp
-cz-us-qwertz		cz,us	pc105		-		terminate:ctrl_alt_bksp,grp:shifts_toggle,grp_led:scroll
-cz-qwerty		cz,us	pc105		qwerty,		terminate:ctrl_alt_bksp,grp:shifts_toggle,grp_led:scroll
 br-abnt2		br	abnt2		-		terminate:ctrl_alt_bksp
-ro			ro	pc105		-		terminate:ctrl_alt_bksp
-us-acentos		us	pc105		intl		terminate:ctrl_alt_bksp
-pt-latin1		pt	pc105		-		terminate:ctrl_alt_bksp
-ro-std-cedilla		ro	pc105		std_cedilla	terminate:ctrl_alt_bksp
-tj_alt-UTF8		tj	pc105		-		terminate:ctrl_alt_bksp
+by			by,us	pc105		-		terminate:ctrl_alt_bksp,grp:shifts_toggle,grp_led:scroll
+cf			ca	pc105		-		terminate:ctrl_alt_bksp
+croat			hr	pc105		-		terminate:ctrl_alt_bksp
+cz-lat2			cz	pc105		qwerty		terminate:ctrl_alt_bksp
+cz-qwerty		cz,us	pc105		qwerty,		terminate:ctrl_alt_bksp,grp:shifts_toggle,grp_led:scroll
+cz-us-qwertz		cz,us	pc105		-		terminate:ctrl_alt_bksp,grp:shifts_toggle,grp_led:scroll
+de			de	pc105		-		terminate:ctrl_alt_bksp
+de-latin1		de	pc105		-		terminate:ctrl_alt_bksp
 de-latin1-nodeadkeys	de	pc105		nodeadkeys	terminate:ctrl_alt_bksp
-no			no	pc105		-		terminate:ctrl_alt_bksp
-bg_bds-utf8		bg,us	pc105		-		terminate:ctrl_alt_bksp,grp:shifts_toggle,grp_led:scroll
-dvorak			us	pc105		dvorak		terminate:ctrl_alt_bksp
+dk			dk	pc105		-		terminate:ctrl_alt_bksp
+dk-latin1		dk	pc105		-		terminate:ctrl_alt_bksp
 dvorak			us	pc105		dvorak-alt-intl	terminate:ctrl_alt_bksp
-ru			ru,us	pc105		-		terminate:ctrl_alt_bksp,grp:shifts_toggle,grp_led:scroll
-cz-lat2			cz	pc105		qwerty		terminate:ctrl_alt_bksp
-pl2			pl	pc105		-		terminate:ctrl_alt_bksp
+dvorak			us	pc105		dvorak		terminate:ctrl_alt_bksp
```

Comment:
There are two internal users of this file. `vconsole_convert_to_x11()` is easy: it matches on the first field, so generally it's not sensitive to sorting. The only exception is `dvorak`, which appears twice, and for which your patch exchanges the two lines, and we'd get a different result.

There's also `find_legacy_keymap()` which does very complicated matching, and I wouldn't want to change the order, unless you can prove that the result would be unchanged (or better).

There are also other users: `calamares` and `guix`. The former does what we do, and the latter just enumerates files to be installed so it is insensitive to order.

Reply:
I didn't expect that the code was sensitive to sorting. If the entries in this file really have to be kept in a certain order, we should put a comment in the file header.

Regarding `find_legacy_keymap()`, it basically finds the best vc layout based on the properties of the passed x11 layout. IMHO such algo should be stable and shouldn't depend on the order of the list.

The problem here is that the code doesn't handle specifically the case when a xvariant is specified but it's not found in the list. It picks up the first matching xlayout it found even if the xvariant doesn't match. It might be better to choose the xlayout *without* an xvariant in this case.

Reply:
Yes, but those things are fragile and we generally don't touch them except to fix a bug.

Reply:
Why not ? it seems to me that "legacy" term used here is a bit misleading.  IUC "legacy" keymaps are those which have not been converted from a x11 layout. If this is the case then converted keymaps can't replace all legacy ones because the latter can include several layouts in the same keymap which is not the case for converted keymaps that include one single layout.

Reply:
I 100% agree with @keszybz here. this is fragile ancient rune stuff; don't touch it just to be tidy. the only reason to touch any of this stuff is to fix a demonstrable bug.

It would definitely be a good idea to add a comment explaining the stuff discussed here though (the users of the file, and that it's sensitive to ordering).

Reply:
> It wouldn't be too hard to verify that after sorting we still get the same results I think.
> 
> I agree that sorting won't benefit to users but it seems to me that it at least shows that this part of the code might be fragile.

well, the logic for picking a layout is pretty complex, so it might be hard to be sure we check *every possible query* where two lines can score the same. for a flavor of the kinds of mess this stuff causes, https://bugzilla.redhat.com/show_bug.cgi?id=2121106#c17 was the last bug I remember fixing in it...

Reply:
Thanks for the interesting pointer.

Side note: commit 950aeeb8ffc950637fac22cb5a42074f227d57f3 appended a comma to the 'xvariant' string. I don't see anything in `find_legacy_keymap()` that would treat ',' specifically. So I guess this commit likely reintroduced the bug that 97f99506980d92e858dc4685a2e84d9548d6eca4 fixed.

Reply:
Ah no sorry ignore my previous comment, I forgot that `xc->xvariant` should also contain the comma hence adding it in kbd-model-map makes sense.

Reply:
Maybe adding a test case for this RH bug would make sense:

```
diff --git a/src/locale/test-localed-util.c b/src/locale/test-localed-util.c
index f57c9ba1f8..c3ee9a9a5c 100644
--- a/src/locale/test-localed-util.c
+++ b/src/locale/test-localed-util.c
@@ -170,6 +170,13 @@ TEST(x11_convert_to_vconsole) {
         assert_se(free_and_strdup(&xc.variant, NULL) >= 0);
         assert_se(x11_convert_to_vconsole(&xc, &vc) >= 0);
         assert_se(streq(vc.keymap, "ru"));
+
+        /* https://bugzilla.redhat.com/show_bug.cgi?id=2121106 */
+        log_info("/* test wit a compount mapping with a known variant (cz,us:qwerty,)");
+        assert_se(free_and_strdup(&xc.layout, "cz,us") >= 0);
+        assert_se(free_and_strdup(&xc.variant, "qwerty,") >= 0);
+        assert_se(x11_convert_to_vconsole(&xc, &vc) >= 0);
+        assert_se(streq(vc.keymap, "cz-qwerty"));
 }
 
 static int intro(void) {
```